examples/dynamoathenamigration/migration.go

```diff
@@ -486,18 +489,251 @@ func (t *task) downloadFromS3(ctx context.Context, key string) (*os.File, error)
 	}
 	path := path.Join(dir, originalName)
 
-	f, err := os.Create(path)
+	originalFile, err := os.Create(path)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	if _, err := t.s3Downloader.Download(ctx, f, &s3.GetObjectInput{
+	if _, err := t.s3Downloader.Download(ctx, originalFile, &s3.GetObjectInput{
 		Bucket: aws.String(t.Bucket),
 		Key:    aws.String(key),
 	}); err != nil {
-		f.Close()
+		originalFile.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	defer originalFile.Close()
+
+	// maxMemoryUsedForSortingExportInBytes defines how big chunks of audit events
+	// will be loaded into memory, sorted and stored in temporary files.
+	maxMemoryUsedForSortingExportInBytes := 1024 * 1024 * t.MaxMemoryUsedForSortingExportInMB
+	sortedFile, err := createSortedExport(originalFile, dir, maxMemoryUsedForSortingExportInBytes)
+	if err != nil {
+		return nil, trace.NewAggregate(err, sortedFile.Close())
+	}
+	return sortedFile, nil
+}
+
+type eventWithTime struct {
+	rawLine   json.RawMessage
+	eventDate string
+}
+
+// createSortedExport reads a large json.gz export file, which cannot fit entirely
+// into memory and splits it into multiple smaller sorted (in memory) files.
+// Afterwards, it merges these smaller files into a final sorted output.
+// The function returns json file which is sorted by createdAtDate ascending.
+func createSortedExport(inFile *os.File, dir string, maxSize int) (f *os.File, err error) {
+	tmpSortedFilesDir := path.Join(dir, fmt.Sprintf("sort-%s", uuid.NewString()))
+	if err := os.Mkdir(tmpSortedFilesDir, 0o755); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer func() {
+		err = trace.NewAggregate(err, os.RemoveAll(tmpSortedFilesDir))
+	}()
+
+	gzipReader, err := gzip.NewReader(inFile)
+	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	return f, nil
+	defer gzipReader.Close()
+
+	dec := json.NewDecoder(gzipReader)
+	events := make([]eventWithTime, 0, 10000)
+	var tmpFiles []string
+	var size int
+	// dec.More read export line by line because export is new line json.
+	for dec.More() {
+		// First decode whole line to rawMessage.
+		var singleLine json.RawMessage
+		err = dec.Decode(&singleLine)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		// Parse just createAtDate.
+		var parsedLine dynamoEventPart
+		err = json.Unmarshal(singleLine, &parsedLine)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		events = append(events, eventWithTime{
+			rawLine:   singleLine,
+			eventDate: parsedLine.Item.CreatedAtDate.Value,
+		})
+		size += len(singleLine)
+		if size >= maxSize {
+			// When max size is reached, sort events and write tmp file.
+			fname, err := sortEventsAndWriteTmpFile(tmpSortedFilesDir, events)
```

Comment:
We can send it off to a separate goroutine. The whole process waits for one file to be sorted and written, but next files don't depend on the result. We can use wait group to sync it and wait for all files to be finished writing and then merge everything.

Reply:
Although it will violate `maxSize` if we do it just like that. But we can something like `chunkSize = maxSize / 4` and run up to 4 sortings in parallel (use `errGroup`?)  limited by `chunkSize` 🤔

Reply:
i don't think this is necessary, most time in migration is spent on generating export from dynamo, which typically takes 5min, sorting time is way fast comparing to generating export. And I am not sure that we will gain any speed boost here, there will be 4x more files. I would prefer to keep it the way it is right now.